server/sonar-server/src/main/java/org/sonar/server/notification/DefaultNotificationManager.java

```diff
@@ -112,54 +112,93 @@ void logDeserializationIssue() {
+        .forEach(subscriberAndChannel -> builder.put(subscriberAndChannel.getSubscriber().getLogin(), subscriberAndChannel.getChannel()));
     }
+    return builder.build();
+  }
 
-    return recipients;
+  private Stream<SubscriberAndChannel> toSubscriberAndChannels(String dispatcherKey, String projectUuid, NotificationChannel notificationChannel) {
+    return dbClient.propertiesDao().findUsersForNotification(dispatcherKey, notificationChannel.getKey(), projectUuid)
+      .stream()
+      .map(login -> new SubscriberAndChannel(login, notificationChannel));
   }
 
-  @Override
-  public Multimap<String, NotificationChannel> findNotificationSubscribers(NotificationDispatcher dispatcher, @Nullable String componentKey) {
-    String dispatcherKey = dispatcher.getKey();
+  private Set<String> keepAuthorizedLogins(String projectUuid, Set<SubscriberAndChannel> subscriberAndChannels, DbSession dbSession) {
+    Set<String> logins = subscriberAndChannels.stream()
+      .map(DefaultNotificationManager.SubscriberAndChannel::getSubscriber)
+      .map(Subscriber::getLogin)
+      .collect(Collectors.toSet());
+    return dbClient.authorizationDao().keepAuthorizedLoginsOnProject(dbSession, logins, projectUuid, UserRole.USER);
+  }
 
-    SetMultimap<String, NotificationChannel> recipients = HashMultimap.create();
-    for (NotificationChannel channel : notificationChannels) {
-      addUsersToRecipientListForChannel(propertiesDao.selectNotificationSubscribers(dispatcherKey, channel.getKey(), componentKey), recipients, channel);
+  private final class SubscriberAndChannel {
+    private final Subscriber subscriber;
+    private final NotificationChannel channel;
+
+    private SubscriberAndChannel(Subscriber subscriber, NotificationChannel channel) {
+      this.subscriber = subscriber;
+      this.channel = channel;
+    }
+
+    public Subscriber getSubscriber() {
+      return subscriber;
     }
 
-    return recipients;
+    public NotificationChannel getChannel() {
+      return channel;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      SubscriberAndChannel that = (SubscriberAndChannel) o;
+      return Objects.equals(subscriber, that.subscriber) &&
+        Objects.equals(channel, that.channel);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(subscriber, channel);
+    }
   }
 
   @VisibleForTesting
   protected List<NotificationChannel> getChannels() {
     return Arrays.asList(notificationChannels);
   }
 
-  private static void addUsersToRecipientListForChannel(List<String> users, SetMultimap<String, NotificationChannel> recipients, NotificationChannel channel) {
+  private static void addUsersToRecipientListForChannel(Collection<String> users, SetMultimap<String, NotificationChannel> recipients, NotificationChannel channel) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this unused private "addUsersToRecipientListForChannel" method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)